### DISCUSSION TITLE
2022-05-13 Developer Tools Update

### DISCUSSION THREAD

**Discussion by austinbaggio** [near/community#274] [Developer Tools]
The developer tools team focuses on NEAR's tools, APIs, & SDKs for both Rust and Javascript. 

## Tools Feedback 
You can reach out to Austin Baggio (Tools Product Manager) to share any feedback about this update or your experience using any of our Dev Tools:  
[30m Calendly Timeslot](https://calendly.com/austinbaggio/30min) | austin.baggio@near.org | [Discord](https://discord.com/channels/490367152054992913/542945453533036544) | [Twitter](https://twitter.com/AustinBaggio)

### near-sdk-js
- @itegulov, @volovyk-s & @ailisp are working on translating SDK-RS examples for the SDK-JS https://github.com/near/near-sdk-js/tree/master/examples
- @volovyk-s Improved the developer onboarding experience by fixing build scripts and READMEs ([PR1](https://github.com/near/near-sdk-js/pull/16))([PR2](https://github.com/near/near-sdk-js/pull/14))
- @ailisp re-implemented nearbindgen to allow recursive @call @view methods
- @ailisp fixed the issue where storage write didn't work correctly after jsvm_call (cross contract call)
- @ailisp Enhanced the on jsvmCall API that encodes arguments and decodes the return for the developer
- @ailisp fixed 6 high level APIs (ecrecover/jsvmStorageRead etc.) that use low-level function returns, BigUint64 (JS BigInt), instead of Number.

### workspaces-rs
- :tada: 0.3.0 version released (includes betanet support and updated protocol changes for sandbox) https://github.com/near/workspaces-rs/pull/136
- @ChaoticTempest started adding error types to allow users to handle specific errors in workspaces-rs
- @ChaoticTempest created the initial draft of what binary version management would look like in both workspaces and sandbox https://hackmd.io/@ChaoticTempest/SkBjQMw85
- @ChaoticTempest exposed raw_bytes API https://github.com/near/workspaces-rs/pull/133
- @ChaoticTempest added more convenient methods to SecretKey type: https://github.com/near/workspaces-rs/pull/139
- @austinabell Updated default sandbox version for workspaces https://github.com/near/workspaces-rs/pull/134
- @austinabell Added check for sandbox initialization to prevent potential bugs https://github.com/near/workspaces-rs/pull/135

### near-cli (JS CLI)
- @ChaoticTempest is scoping the JS CLI commands needed to interact with JavaScript contracts

### ABI 
- @itegulov is working on a Rust ABI codegen library, it can already generate valid contract method cascades but actual RPC calls inside is still WIP

### jsonrpc-client
- @miraclx fixed UnknownBlock variant parsing. https://github.com/near/near-jsonrpc-client-rs/pull/94
- @miraclx added logging, making the client's request / response payloads inspectable. https://github.com/near/near-jsonrpc-client-rs/pull/97
- @miraclx Extended the create account example showing how to create sub-accounts & top-level accounts. https://github.com/near/near-jsonrpc-client-rs/blob/master/examples/create_account.rs
- @miraclx Reworked the JsonRpcError::handler_error method. https://github.com/near/near-jsonrpc-client-rs/pull/99

### nearcore
- @miraclx is working through dependencies while updating actix-web to stable. It needs more investigation into migration changeshttps://github.com/near/nearcore/pull/6694
- @miraclx is working on releasing a new version (0.13.0) of public nearcore crates.

### near-sdk-rs
- @austinabell 
Release 4.0.0-pre.9 of SDK https://github.com/near/near-sdk-rs/pull/814
- Updated cross-contract call API change merged https://github.com/near/near-sdk-rs/pull/742

**Comment by miraclx** [comment]
@thisisjoshford 

> - fixed `UnknownBlock` variant parsing. https://github.com/near/near-jsonrpc-client-rs/pull/94

The context behind this is, since we reuse the same structures as the server, we expect to be able to deserialize whatever the server serialized.

However, there are some cases where the server skips the serialization of some fields (https://github.com/near/near-jsonrpc-client-rs/issues/28). Making it impossible to deserialize on the client side.

A patch was applied (https://github.com/near/near-jsonrpc-client-rs/pull/29) to mitigate that, but that incorrectly treats the legacy, string-typed `.data` field as a JSON object, instead of the more appropriate `.error_struct` field from the server response.
